music21/spanner.py

```diff
@@ -610,6 +705,84 @@ class _SpannerRef(t.TypedDict):
     spanner: 'Spanner'
     className: str
 
+class SpannerAnchor(base.Music21Object):
+    '''
+    A simple Music21Object that can be used to define the beginning or end
+    of a Spanner, in the place of a GeneralNote.
+
+    This is useful for (e.g.) a Crescendo that ends partway through a
+    note (e.g. in a violin part).
+
+    Here's an example of a whole note that has a Crescendo for the first
+    half of the note, and a Diminuendo for the second half of the note.
+
+    >>> score = stream.Score()
+    >>> part = stream.Part()
+    >>> score.insert(0, part)
+    >>> measure = stream.Measure()
+    >>> part.insert(0, measure)
+    >>> voice = stream.Voice()
+    >>> measure.insert(0, voice)
+
+    >>> n = note.Note('C', quarterLength = 4)
+    >>> sa1 = spanner.SpannerAnchor()
+    >>> sa2 = spanner.SpannerAnchor()
+    >>> voice.insert(0, n)
+    >>> voice.insert(2, sa1)
+    >>> voice.insert(4, sa2)
+    >>> cresc = dynamics.Crescendo(n, sa1)   # cresc from n to sa1
+    >>> dim = dynamics.Diminuendo(sa1, sa2)  # dim from sa1 to sa2
+    >>> score.append((cresc, dim))
+    >>> score.show('text')
+    {0.0} <music21.stream.Part 0x...>
+        {0.0} <music21.stream.Measure 0 offset=0.0>
+            {0.0} <music21.stream.Voice 0x...>
+                {0.0} <music21.note.Note C>
+                {2.0} <music21.spanner.SpannerAnchor object at 0x...>
+                {4.0} <music21.spanner.SpannerAnchor object at 0x...>
+    {4.0} <music21.dynamics.Crescendo <music21.note.Note C><music21.spanner.SpannerAnchor...>>
+    {4.0} <music21.dynamics.Diminuendo <...SpannerAnchor...><...SpannerAnchor...>>
+
+    SpannerAnchors aways have a duration of 0, and if any attempt is made to
+    change this, TypeError will be raised.
```

Comment:
Hmmm... I think of a SpannerAnchor as anchoring one end of the spanner, never both.  However, your point about things like augmentOrDiminish is compelling.  I will allow SpannerAnchors to have duration, and if the SpannerAnchor is the last (or only) element in a Spanner, then m21ToXml.py will need to compute the spanner end offset as anchor.offset + anchor.duration.

Reply:
Looks like m21ToXML.py already gets it right by moving forward when it skips the element (because that element might be a SpannerAnchor, or it might be a Note already emitted in a previous pass):

```
        else:  # appendSpanners == AppendSpanners.RELATED_ONLY
            # skip the element itself, just move forward by element duration
            if postList and obj.quarterLength > 0:
```

Reply:
done.